app/services/atmosphere/cloud/satisfy_appliance.rb

```diff
@@ -11,8 +11,16 @@ def execute
         if appliance.virtual_machines.blank?
           if optimization_strategy.can_reuse_vm? &&
             !(vm = optimization_strategy.vm_to_reuse).nil?
-
             action.log(I18n.t('satisfy_appliance.reusing', vm: vm.id_at_site))
+            # Assign correct fund for this VM to appliance
+            cfs = vm.tenant.funds & appliance.appliance_set.user.funds
+            if cfs.include? appliance.appliance_set.user.default_fund
+              appliance.fund = appliance.appliance_set.user.default_fund
+              appliance.save
+            elsif cfs.length > 0
+              appliance.fund = cfs.first
+              appliance.save
+            end
```

Comment:
Please extract this code into separate method (now we have different level of abstraction code in one method). You can use following construction to assign `fund` only when `nil`.

```ruby
appliance.fund ||= select_fund(vm)
```